src/wifipaf/WiFiPAFEndPoint.cpp

```diff
@@ -0,0 +1,1255 @@
+    {
+        Free();
+    }
+    else // Otherwise, try to signal close to remote device before end point releases WiFiPAF connection and frees itself.
+    {
+        if (mRole == kWiFiPafRole_Subscriber)
+        {
+            // Cancel send and receive-ack timers, if running.
+            StopAckReceivedTimer();
+            StopSendAckTimer();
+            mConnStateFlags.Set(ConnectionStateFlag::kOperationInFlight);
+        }
+        else
+        {
+            Free();
+        }
+    }
+    ClearAll();
+}
+
+void WiFiPAFEndPoint::DoCloseCallback(uint8_t state, uint8_t flags, CHIP_ERROR err)
+{
+    // Callback fires once per end point lifetime.
+    mOnPafSubscribeComplete = nullptr;
+    mOnPafSubscribeError    = nullptr;
+    OnConnectionClosed      = nullptr;
+}
+
+void WiFiPAFEndPoint::Free()
+{
+    // Clear fragmentation and reassembly engine's Tx and Rx buffers. Counters will be reset by next engine init.
+    FreePAFtpEngine();
+
+    // Clear pending ack buffer, if any.
+    mAckToSend = nullptr;
+
+    // Cancel all timers.
+    StopConnectTimer();
+    StopAckReceivedTimer();
+    StopSendAckTimer();
+
+    // Clear callbacks.
+    mOnPafSubscribeComplete = nullptr;
+    mOnPafSubscribeError    = nullptr;
+    OnMessageReceived       = nullptr;
+    OnConnectionClosed      = nullptr;
+}
+
+void WiFiPAFEndPoint::FreePAFtpEngine()
+{
+    // Free transmit disassembly buffer
+    mPafTP.ClearTxPacket();
+
+    // Free receive reassembly buffer
+    mPafTP.ClearRxPacket();
+}
+
+CHIP_ERROR WiFiPAFEndPoint::Init(WiFiPAFLayer * WiFiPafLayer, WiFiPAFSession & SessionInfo)
+{
+    // Fail if already initialized.
+    VerifyOrReturnError(mWiFiPafLayer == nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    // Validate args.
+    VerifyOrReturnError(WiFiPafLayer != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+
+    // If end point plays subscriber role, expect ack as last step of PAFTP handshake.
+    // If being publisher, subscriber's handshake indication 'ack's write sent by publisher to kick off the PAFTP handshake.
+    bool expectInitialAck = (SessionInfo.role == kWiFiPafRole_Publisher);
+
+    CHIP_ERROR err = mPafTP.Init(this, expectInitialAck);
+    if (err != CHIP_NO_ERROR)
+    {
+        ChipLogError(WiFiPAF, "WiFiPAFTP init failed");
+        return err;
+    }
+
+    mWiFiPafLayer = WiFiPafLayer;
+
+    // WiFiPAF EndPoint data members:
+    memcpy(&mSessionInfo, &SessionInfo, sizeof(mSessionInfo));
+    mRole = SessionInfo.role;
+    mTimerStateFlags.ClearAll();
+    mLocalReceiveWindowSize  = 0;
+    mRemoteReceiveWindowSize = 0;
+    mReceiveWindowMaxSize    = 0;
+    mSendQueue               = nullptr;
+    mAckToSend               = nullptr;
+
+    ChipLogDebugWiFiPAFEndPoint(WiFiPAF, "initialized local rx window, size = %u", mLocalReceiveWindowSize);
+
+    // End point is ready.
+    mState = kState_Ready;
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR WiFiPAFEndPoint::SendCharacteristic(PacketBufferHandle && buf)
```

Comment:
Is it called a "characteristic" for PAF?  BLE has "characteristics", sure, but I find nothing called "characteristic" for PAF....

Reply:
It should be the follow-up packets. I will change it